### ISSUE TITLE
Incorrect handling of .post0 modifiers in comparisons

### ISSUE THREAD

**Issue by njsmith** [relrod/pep440-rs#1]
Well, this is rather nitpicky :-)

According to PEP 440:

> The post-release segment consists of the string .post, followed by a non-negative integer value. Post-releases are ordered by their numerical component, immediately following the corresponding release

"non-negative integer" means that `.post0` is a valid post-release segment. And since it's a post-release segment, it should be ordered after the corresponding release. This also matches the behavior of `packaging`:

```python
>>> from packaging.version import Version                                             
>>> Version("1.0") == Version("1.0.post0")                                            
False
>>> Version("1.0") < Version("1.0.post0")                                             
True
```

However, the code in this crate treats those two versions as identical for purposes of comparison: `==` returns True, `<` returns False.

I believe the bug is here:

https://github.com/relrod/pep440-rs/blob/f031645e4ccec11f342fe3f593a41cfaf8402ee5/src/lib.rs#L443

`post: None` and `post: Some(0)` both get mapped to `0` for comparison purposes.

(Discovered this because I was messing with using https://crates.io/crates/pubgrub to resolve python package versions, and that algorithm wants to be able to represent version constraints as ranges. So `== 1.0` compiles down to the range `[1.0, 1.0post0)`.)

**Comment by relrod** [comment]
> Well, this is rather nitpicky :-)

Nope, it's a valid bug and I appreciate the report :)

I'll try to get it patched up soon. This is one of my first Rust projects, so definitely open to any other critique/bug reports as well. I know the code is a bit ugly in some places right now.

**Comment by njsmith** [comment]
No worries, I have no idea what I'm doing in rust either :-)

I did notice that `packaging` has a much simpler way of implementing comparisons -- it's a *bit* awkward to convert to rust because it makes free use of mixed type comparisons like `NegativeInfinity < 0`, but I think with a bit of work it could be done and might be easier to work with than the current code?

https://github.com/pypa/packaging/blob/6b2efc63e3723f7af2d914e656e93cdcc15b6c68/packaging/version.py#L444

Since it works by generating a kind of normalized key, it would also be useful for implementing `Hash` for `pep440::Version`, which is something I needed to tack on. Currently I'm feeding `version.normalize()` into the hash function, which I think probably ought to work, but it's definitely more expensive and fragile than feeding in the comparison data directly.
